### ISSUE TITLE
performance question regarding file size and splitting

### ISSUE THREAD

**Issue by ziegi** [animetosho/par2cmdline-turbo#28]
Thank you for coding this fast par2 alternative.

As I am working with large files and expect a fixed corruption length of 2MB I have to split my input files to not exceed the 32768 source block count.

I found that par2cmdline-turbo is working faster on multiple small files than on one big file.
Increasing Memory does not accelerate  the calculation of big files

But I do not understand what is the cause for this or the implications of splitting files for better performance,
as in my understanding the results and error correction possibilities should be the same.

common parameters for all tests: 
par2cmdline-turbo v1.1.1
c -s2097152 -r10 -u -m16777216 -t7
number of output files are adjusted from -n10 to -n20 and -n40, -n80, so each hold 81 or 82 recovery blocks of 2MB
to have same amount of files created
time for splitted files of course include splitting process

Test Code example:
`s=8G; n=4; f=test_32G.bin; sync ; echo 3 | sudo tee -a /proc/sys/vm/drop_caches; time ../par2 c -s2097152 -r10 -u -n$((n*10)) -m16777216 -t7 -vv $f; sync ; echo 3 | sudo tee -a /proc/sys/vm/drop_caches; time (split -b $s -d $f $f"-" && for i in $f"-"*; do ../par2 c -s2097152 -r10 -u -n10 -m16777216 -t7 -vv $i; done);`

Test results: Filsize : Time in seconds
```
                                        16GB :  49s  vs  2x 8GB : 49s  vs  4x 4GB : 44s
                    32GB : 185s  vs  2x 16GB : 142s  vs  4x 8GB : 105s
64GB : 716s  vs  2x 32GB : 447s  vs  4x 16GB : 278s  vs  8x 8GB : 203s
```

Am I Missing something ?

**Comment by animetosho** [comment]
The time complexity for PAR2 computation is proportional to `input size * number of recovery blocks`.

Let's take your 32GB vs 2x16GB example.

* 32GB input file: recovery blocks = 32GB\*10% / 2MB = 1638, time complexity = 32GB \* 1638 = 51.19TB
* 16GB input file: recovery blocks = 16GB\*10% / 2MB = 819, time complexity = 16GB \* 819 = 12.80TB

As there's two files for the second case, we double it to arrive at 25.6TB, so one would expect it'd run about twice as fast (in reality, scaling isn't perfect, so it'd be a bit less, plus you're timing the split).

> as in my understanding the results and error correction possibilities should be the same.

They aren't quite the same.

In your 1x32GB example, the PAR2 can recover up to 1638 blocks anywhere in the file.  
For the 16GB example, the PAR2 can only recovery up to 819 blocks in that file. Since there's two files, you might think that 2\*819 gives you the same 1638 blocks of redundancy, but the errors need to be divided such that no more than 819 blocks of corruption occur in each file.  
In other words, if 1638 blocks are corrupt at the beginning of the 32GB file, it's recoverable, but if that number is corrupt at the beginning of a 16GB file, it's not recoverable.

> Increasing Memory does not accelerate the calculation of big files

The memory is used to hold recovery data as it's being computed. For example, with 32GB input at 10% recovery, you're computing 3.2GB recovery data, so increasing memory limit beyond that point doesn't really do anything.  
There is also diminishing returns for larger memory limits anyway.

**Comment by ziegi** [comment]
Thanks a lot for the quick explanation.

I see the difference now, that the largest possible 
corruption is only half as long.By splitting the file in two.

Regarding my use case (errors on magnetic tape)
I know that usually the errors are distributed across the whole file
and not concentrated in one spot.
So I guess I am fine with splitting the file, 
trading encryption time against corruption length.
